<a href="https://colab.research.google.com/github/argenviahouse-sys/Procesamiento-del-Habla/blob/main/TP3_Embedding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# TP 3 - Embeddings

Asignatura, Profesor: Procesamiento del Habla, Ana Laura Diedrich

Alumno, Carrera: Ivan David Mella, IA

Consigna de trabajo: análisis de similitud semántica con spacy

Objetivo: Explorar y comprender las relaciones semánticas entre palabras mediante el cálculo de la similitud del coseno utilizando modelos de lenguaje.

Instrucciones:

Cargar y utilizar los embeddings en español de la librería spaCy.

Mediante código, identificar tres pares de palabras que cumplan con las siguientes condiciones de similitud del coseno:

Similitud positiva: un par de palabras con un valor cercano a 1.

Similitud neutra: un par de palabras con un valor cercano a 0.

Similitud negativa: un par de palabras con un valor cercano a -1 (o el valor más negativo que logres encontrar).

Interpretar los resultados: redactar una explicación analizando por qué cada par de palabras arrojó ese nivel de similitud según su significado y uso en el lenguaje.

In [1]:
# Instalación de spaCy y descarga del modelo en español
!pip install spacy --quiet
!python -m spacy download es_core_news_sm --quiet

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.9/12.9 MB 82.9 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [4]:
import spacy

# Carga del modelo en español
nlp = spacy.load("es_core_news_sm")


In [5]:
# Función para calcular la similitud
def medir_similitud(p1, p2):
    token1 = nlp(p1)
    token2 = nlp(p2)
    return token1.similarity(token2)

# Pares de palabras según la consigna
# 1. Similitud Positiva (~1)
par_positivo = ("feliz", "contento")

# 2. Similitud Neutra (~0)
par_neutro = ("algoritmo", "arena")

# 3. Similitud Negativa (Mínima posible)
par_negativo = ("fuego", "hielo")

# Impresión de resultados
print(f"Similitud Positiva (Sinónimos): {par_positivo} -> {medir_similitud(*par_positivo):.4f}")
print(f"Similitud Neutra (Sin relación): {par_neutro} -> {medir_similitud(*par_neutro):.4f}")
print(f"Similitud Mínima (Opuestos): {par_negativo} -> {medir_similitud(*par_negativo):.4f}")

Similitud Positiva (Sinónimos): ('feliz', 'contento') -> 0.3689
Similitud Neutra (Sin relación): ('algoritmo', 'arena') -> 0.2422
Similitud Mínima (Opuestos): ('fuego', 'hielo') -> 0.4407


/tmp/ipykernel_2335/2916475876.py:5: UserWarning: [W007] The model you're using has no word vectors loaded, so the result of the Doc.similarity method will be based on the tagger, parser and NER, which may not give useful similarity judgements. This may happen if you're using one of the small models, e.g. `en_core_web_sm`, which don't ship with word vectors and only use context-sensitive tensors. You can always add your own word vectors, or use one of the larger models instead if available.
  return token1.similarity(token2)


El sistema nos advierte que el modelo SM usado no es grande y se deberia probar con uno mediano:
ya que la limitación de los modelos "sm" es que para ser ligeros y rápidos, estos modelos no incluyen una tabla de vectores preentrenado.

Si el modelo no tiene vectores cargados, los tensores que utiliza son "sensibles al contexto" pero no representan el significado profundo de la palabra de forma independiente.
Sus resultados de similitud no se basan en el significado profundo de la palabra, sino en estimaciones del contexto gramatical y sintáctico (tagger y parser)

El modelo es_core_news_sm utiliza tensores sensibles al contexto pero no incluye vectores de palabras preentrenados. Al usar el método .similarity(), spaCy emite una advertencia indicando que los resultados pueden no ser útiles.

**Solucion:** Usar un modelo más grande en lugar de es_core_news_sm, se debería descargar y cargar la versión mediana o grande, que sí "embarca" vectores de palabras reales.

El modelo MD (Medium) es significativamente más preciso para tareas de similitud semántica.

Este modelo sí "embarca" word vectors preentrenados, lo que le permite situar las palabras en un mapa matemático de 300 dimensiones donde la cercanía física representa similitud de significado.


In [1]:
!python -m spacy download es_core_news_md

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.3/42.3 MB 16.6 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_md')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [2]:
import spacy

# Carga del modelo en español
nlp2 = spacy.load("es_core_news_md")

In [3]:
# Función para calcular la similitud
def medir_similitud(p1, p2):
    token1 = nlp2(p1)
    token2 = nlp2(p2)
    return token1.similarity(token2)

# Pares de palabras según la consigna
# 1. Similitud Positiva (~1)
par_positivo = ("feliz", "contento")

# 2. Similitud Neutra (~0)
par_neutro = ("algoritmo", "arena")

# 3. Similitud Negativa (Mínima posible)
par_negativo = ("fuego", "hielo")

# Impresión de resultados
print(f"Similitud Positiva (Sinónimos): {par_positivo} -> {medir_similitud(*par_positivo):.4f}")
print(f"Similitud Neutra (Sin relación): {par_neutro} -> {medir_similitud(*par_neutro):.4f}")
print(f"Similitud Mínima (Opuestos): {par_negativo} -> {medir_similitud(*par_negativo):.4f}")

Similitud Positiva (Sinónimos): ('feliz', 'contento') -> 0.3892
Similitud Neutra (Sin relación): ('algoritmo', 'arena') -> 0.0356
Similitud Mínima (Opuestos): ('fuego', 'hielo') -> 0.4586


# Interpretación de los resultados específicos
**Similitud Positiva** (feliz / contento):
SM (0.3689) vs. MD (0.3892): Ambos detectan una relación, pero el modelo MD es ligeramente más alto. Al ser sinónimos, se espera que aparezcan en contextos casi idénticos, lo que acerca sus vectores
. El valor del modelo MD es más robusto porque proviene de una representación densa de significado.

**Similitud Neutra** (algoritmo / arena):
SM (0.2422) vs. MD (0.0356): Esta es la diferencia más reveladora. El modelo SM otorga un 0.24 (una relación moderada inexistente), lo cual es "ruido" técnico [User Warning]. En cambio, el modelo MD arroja un 0.03, un valor casi nulo que refleja correctamente que no existe ninguna relación semántica ni contextual entre un término informático y uno natural.

**Similitud Mínima/Opuestos** (fuego / hielo):
SM (0.4407) vs. MD (0.4586): Aquí ambos modelos dan valores relativamente altos (cercanos a 0.5). Esto ocurre porque, aunque son opuestos, son contextualmente similares: ambos describen temperaturas extremas y aparecen en oraciones similares (ej. "sentir el..."). Las fuentes explican que los antónimos suelen tener similitudes positivas porque comparten el mismo "espacio" de uso en el lenguaje
el modelo MD es el adecuado, ya que sus puntuaciones discriminan mucho mejor los términos no relacionados (como se vio en el par algoritmo/arena) y validan las relaciones semánticas basándose en vectores densos y no solo en reglas gramaticales

Ahora se probara con un modelo grande
para comparar los resultados de las distancias vectoriales

In [4]:
!python -m spacy download es_core_news_lg

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 568.0/568.0 MB 1.2 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_lg')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [5]:
import spacy

# Carga del modelo en español
nlp3 = spacy.load("es_core_news_lg")

In [27]:
# Función para calcular la similitud
def medir_similitud(p1, p2):
    token1 = nlp3(p1)
    token2 = nlp3(p2)
    return token1.similarity(token2)

# Pares de palabras según la consigna
# 1. Similitud Positiva (~1)
par_positivo = ("feliz", "contento")

# 2. Similitud Neutra (~0)
par_neutro = ("algoritmo", "arena")

# 3. Similitud Negativa (Mínima posible)
par_negativo = ("fuego", "hielo")

# Impresión de resultados
print(f"Similitud Positiva (Sinónimos): {par_positivo} -> {medir_similitud(*par_positivo):.4f}")
print(f"Similitud Neutra (Sin relación): {par_neutro} -> {medir_similitud(*par_neutro):.4f}")
print(f"Similitud Mínima (Opuestos): {par_negativo} -> {medir_similitud(*par_negativo):.4f}")

Similitud Positiva (Sinónimos): ('feliz', 'contento') -> 0.3892
Similitud Neutra (Sin relación): ('algoritmo', 'arena') -> 0.0356
Similitud Mínima (Opuestos): ('fuego', 'hielo') -> 0.4586


# Análisis Detallado de tus Resultados
**Similitud Positiva** (feliz / contento) -> 0.3892:
Interpretación: Es significativamente superior al 0.36 del modelo SM. Aunque 0.38 puede parecer un número bajo para sinónimos, en el mundo de los embeddings es un valor sólido. Indica que estas palabras comparten una "dirección" similar en el mapa 3D porque los humanos las usamos en situaciones casi idénticas [Conversación previa, 131].

**Similitud Neutra** (algoritmo / arena) -> 0.0356:
Interpretación: Este es el resultado más exitoso del modelo LG/MD. Mientras que el modelo SM "alucinaba" una relación de 0.24, el modelo LG detecta casi un 0 absoluto. Esto confirma que los vectores son ortogonales (perpendiculares); no hay absolutamente ninguna conexión de significado ni de contexto entre la informática y la geología.

**Similitud Mínima/Opuestos** (fuego / hielo) -> 0.4586:
Interpretación: Paradójicamente, este es el valor más alto. ¿Por qué? Porque los embeddings capturan contexto, no solo definición. "Fuego" y "hielo" aparecen siempre en textos sobre temperatura, sensaciones térmicas o fenómenos físicos
. El modelo LG reconoce que, aunque son opuestos, pertenecen al mismo "mundo" semántico, por lo que sus vectores están más cerca entre sí que "algoritmo" y "arena"

In [12]:
# Función para calcular la similitud
def medir_similitud(p1, p2):
    token1 = nlp3(p1)
    token2 = nlp3(p2)
    return token1.similarity(token2)

# Pares de palabras según la consigna
# 1. Similitud Positiva (~1)
par_positivo = ("superfeliz", "hipercontento")

# 2. Similitud Neutra (~0)
par_neutro = ("algoritmico", "arenero")

# 3. Similitud Negativa (Mínima posible)
par_negativo = ("fuegos", "hielos")

# Impresión de resultados
print(f"Similitud Positiva (Sinónimos): {par_positivo} -> {medir_similitud(*par_positivo):.4f}")
print(f"Similitud Neutra (Sin relación): {par_neutro} -> {medir_similitud(*par_neutro):.4f}")
print(f"Similitud Mínima (Opuestos): {par_negativo} -> {medir_similitud(*par_negativo):.4f}")

Similitud Positiva (Sinónimos): ('superfeliz', 'hipercontento') -> 0.0000
Similitud Neutra (Sin relación): ('algoritmico', 'arenero') -> 0.0000
Similitud Mínima (Opuestos): ('fuegos', 'hielos') -> 0.4664


/tmp/ipykernel_10359/1853514520.py:5: UserWarning: [W008] Evaluating Doc.similarity based on empty vectors.
  return token1.similarity(token2)


# Interpretación de los Resultados (0.0000)
El hecho de que los pares ('superfeliz', 'hipercontento') y ('algoritmico', 'arenero') arrojen una similitud de 0.0000 no significa que no tengan relación, sino que el modelo no conoce esas palabras específicas.

**Vectores de ceros:** Según las fuentes, cuando spaCy encuentra una palabra que no está en su diccionario preentrenado, le asigna un vector vacío (lleno de ceros).

**Matemáticamente:** Al intentar calcular la similitud del coseno con un vector de ceros, el resultado siempre será 0, ya que no hay "dirección" ni "magnitud" que comparar en el espacio vectorial.

Palabras complejas con prefijos como "super-" o "hiper-" a menudo no están indexadas de forma literal en modelos como es_core_news_md o lg.

El caso de ('fuegos', 'hielos') -> 0.4664
A diferencia de los anteriores, este par sí arrojó un resultado válido y muy similar al que obtuviste con sus versiones en singular ("fuego" y "hielo" -> 0.4586).

Palabras conocidas: Esto indica que el modelo sí tiene en su vocabulario las formas plurales "fuegos" e "hielos".
Consistencia semántica: El valor se mantiene cercano al 0.46 porque, aunque cambie el sufijo de número (plural), el contexto de uso (temperatura, elementos físicos) sigue siendo el mismo en el corpus de entrenamiento.

La advertencia UserWarning: [W008] Evaluating Doc.similarity based on empty vectors es la confirmación técnica de lo anterior.
Significado: Está avisando que el cálculo de similitud se está realizando sobre vectores vacíos.

Consecuencia: El sistema te advierte que el resultado de 0.0 no es un juicio semántico real (es decir, no es que las palabras sean "opuestas" o "diferentes"), sino que es un error de datos porque el modelo no tiene información sobre esas palabras para poder compararlas.

Resumen del análisis:
OOV (Out of Vocabulary): Es la gran desventaja de modelos como Word2Vec o spaCy.

Si la palabra no es idéntica a una vista en el entrenamiento, el modelo "se pierde".

Solución técnica: En la práctica industrial, para evitar esto se suelen usar técnicas de lematización previa (llevar "superfeliz" a "feliz") o utilizar modelos como FastText, que pueden entender partes de palabras (subpalabras) y así "adivinar" el vector de una palabra nueva basándose en sus raíces y prefijos